v2/event/event_unmarshal.go

```diff
@@ -0,0 +1,447 @@
+		if err != nil {
+			return err
+		}
+		e.DataEncoded = e.DataEncoded[0:len]
+		return nil
+	}
+
+	ct := e.DataContentType()
+	if ct != ApplicationJSON && ct != TextJSON {
+		// If not json, then data is encoded as string
+		src := iter.ReadStringAsSlice()
+		e.DataEncoded = make([]byte, len(src))
+		copy(e.DataEncoded, src)
+		return nil
+	}
+	e.DataEncoded = iter.SkipAndReturnBytes()
+	return nil
+}
+
+type tokenQueue struct {
+	slice []entry
+	i     int
+}
+
+func (q *tokenQueue) push(key string, raw jsoniter.Any) {
+	if q.i < len(q.slice) {
+		q.slice[q.i].key = key
+		q.slice[q.i].value = raw
+	} else {
+		e := entry{key: key, value: raw}
+		q.slice = append(q.slice, e)
+	}
+
+	q.i++
+}
+
+func readUriRef(iter *jsoniter.Iterator) types.URIRef {
+	str := iter.ReadString()
+	uriRef := types.ParseURIRef(str)
+	if uriRef == nil {
+		iter.Error = fmt.Errorf("cannot parse uri ref: %v", str)
+		return types.URIRef{}
+	}
+	return *uriRef
+}
+
+func readStrPtr(iter *jsoniter.Iterator) *string {
+	str := iter.ReadString()
+	if str == "" {
+		return nil
+	}
+	return &str
+}
+
+func readUriRefPtr(iter *jsoniter.Iterator) *types.URIRef {
+	return types.ParseURIRef(iter.ReadString())
+}
+
+func readUriPtr(iter *jsoniter.Iterator) *types.URI {
+	return types.ParseURI(iter.ReadString())
+}
+
+func readTimestamp(iter *jsoniter.Iterator) *types.Timestamp {
+	t, err := types.ParseTimestamp(iter.ReadString())
+	if err != nil {
+		iter.Error = err
+	}
+	return t
+}
+
+func toUriRef(val jsoniter.Any) (types.URIRef, error) {
+	str := val.ToString()
+	if val.LastError() != nil {
+		return types.URIRef{}, val.LastError()
+	}
+	uriRef := types.ParseURIRef(str)
+	if uriRef == nil {
+		return types.URIRef{}, fmt.Errorf("cannot parse uri ref: %v", str)
+	}
+	return *uriRef, nil
+}
+
+func toStrPtr(val jsoniter.Any) (*string, error) {
+	str := val.ToString()
+	if val.LastError() != nil {
+		return nil, val.LastError()
+	}
+	if str == "" {
+		return nil, nil
+	}
+	return &str, nil
+}
+
+func toUriRefPtr(val jsoniter.Any) (*types.URIRef, error) {
+	str := val.ToString()
+	if val.LastError() != nil {
+		return nil, val.LastError()
+	}
+	return types.ParseURIRef(str), nil
+}
+
+func toUriPtr(val jsoniter.Any) (*types.URI, error) {
+	str := val.ToString()
+	if val.LastError() != nil {
+		return nil, val.LastError()
+	}
+	return types.ParseURI(str), nil
+}
+
+func toTimestamp(val jsoniter.Any) (*types.Timestamp, error) {
+	str := val.ToString()
+	if val.LastError() != nil {
+		return nil, val.LastError()
+	}
+	t, err := types.ParseTimestamp(str)
+	if err != nil {
+		return nil, err
+	}
+	return t, nil
+}
+
+// UnmarshalJSON implements the json unmarshal method used when this type is
+// unmarshaled using json.Unmarshal.
+func (e *Event) UnmarshalJSON(b []byte) error {
+	_, r := observability.NewReporter(context.Background(), eventJSONObserved{o: reportUnmarshal})
+	err := ReadJson(e, bytes.NewReader(b))
+
+	// Report the observable
+	if err != nil {
+		r.Error()
+	} else {
```

Comment:
Rather than an else:

```
// Report the observable
	if err := ReadJson(e, bytes.NewReader(b)); err != nil {
		r.Error()
		return err
	}
	r.OK()
	return nil
```